src/bpf-loader-upgradeable.ts

```diff
@@ -0,0 +1,699 @@
+        {pubkey: SYSVAR_CLOCK_PUBKEY, isSigner: false, isWritable: false},
+        {
+          pubkey: params.authorityPubkey,
+          isSigner: true,
+          isWritable: false,
+        },
+      ],
+      programId: this.programId,
+      data,
+    });
+  }
+
+  /**
+   * Generate a transaction instruction that set a new buffer authority
+   */
+  static setBufferAuthority(
+    params: SetBufferAuthorityParams,
+  ): TransactionInstruction {
+    const type = BPF_UPGRADEABLE_LOADER_INSTRUCTION_LAYOUTS.SetAuthority;
+    const data = encodeData(type, {});
+
+    return new TransactionInstruction({
+      keys: [
+        {pubkey: params.bufferPubkey, isSigner: false, isWritable: true},
+        {
+          pubkey: params.authorityPubkey,
+          isSigner: true,
+          isWritable: false,
+        },
+        {pubkey: params.newAuthorityPubkey, isSigner: false, isWritable: false},
+      ],
+      programId: this.programId,
+      data,
+    });
+  }
+
+  /**
+   * Generate a transaction instruction that set a new program authority
+   */
+  static async setUpgradeAuthority(
+    params: SetUpgradeAuthorityParams,
+  ): Promise<TransactionInstruction> {
+    const type = BPF_UPGRADEABLE_LOADER_INSTRUCTION_LAYOUTS.SetAuthority;
+    const data = encodeData(type, {});
+
+    const programDataPubkey = await this.getProgramDataAddress(
+      params.programPubkey,
+    );
+
+    const keys = [
+      {pubkey: programDataPubkey, isSigner: false, isWritable: true},
+      {
+        pubkey: params.authorityPubkey,
+        isSigner: true,
+        isWritable: false,
+      },
+    ];
+
+    if (params.newAuthorityPubkey) {
+      keys.push({
+        pubkey: params.newAuthorityPubkey,
+        isSigner: false,
+        isWritable: false,
+      });
+    }
+
+    return new TransactionInstruction({
+      keys,
+      programId: this.programId,
+      data,
+    });
+  }
+
+  /**
+   * Generate a transaction instruction that close program, buffer, or
+   *   uninitialized account
+   */
+  static close(params: CloseParams): TransactionInstruction {
+    const type = BPF_UPGRADEABLE_LOADER_INSTRUCTION_LAYOUTS.Close;
+    const data = encodeData(type, {});
+
+    const keys = [
+      {pubkey: params.closePubkey, isSigner: false, isWritable: true},
+      {
+        pubkey: params.recipientPubkey,
+        isSigner: false,
+        isWritable: true,
+      },
+    ];
+
+    if (params.authorityPubkey) {
+      keys.push({
+        pubkey: params.authorityPubkey,
+        isSigner: true,
+        isWritable: false,
+      });
+    }
+
+    if (params.programPubkey) {
+      keys.push({
+        pubkey: params.programPubkey,
+        isSigner: false,
+        isWritable: true,
+      });
+    }
+
+    return new TransactionInstruction({
+      keys,
+      programId: this.programId,
+      data,
+    });
+  }
+}
+
+/**
+ * BpfLoaderUpgradeable program interface
+ */
+export class BpfLoaderUpgradeable {
```

Comment:
I would prefer to not include this class and all of its methods inside web3. I think they could live inside a separate package if needed but I don't think it's too valuable to deploy programs using JS right now. It's a complex process to ensure that all transactions get processed properly and I think that it's fine to keep that process only in the CLI for now. It's too complex to maintain in web3.js for now.

Reply:
:confused: is it possible to store and maintain a separated package with `BpfLoaderUpgradeable` under `solana-labs` organization or I would need to store and distribute under my own account?

Reply:
Separate account for now until we setup a place for solana community web3 packages. Sorry about that! Maintenance can be a burden 😉